dcmgr/lib/dcmgr/models/network.rb

```diff
@@ -154,45 +154,137 @@ def ipv4_u32_dynamic_range_array
-           range.destroy
-         end
-         range.save_changes
-      }
+      range_begin_u32, range_end_u32 = validate_range_args(range_begin, range_end)
+      dhcp_range_dataset.included_ranges(range_begin_u32, range_end_u32).destroy
+      hit_range_ds = dhcp_range_dataset.hit_ranges(range_begin_u32, range_end_u32)
+      if hit_range_ds.empty?
+        # no overwrap ranges. add new range.
+        return self.add_dhcp_range(range_begin: range_begin_u32,
+                                   range_end: range_end_u32)
+      end
 
-      self.add_dhcp_range(:range_begin=>range_begin, :range_end=>range_end)
+      to_be_merged=[nil, nil]
+      hit_range_ds.each { |r|
+        existing_range = (r.range_begin.to_u32 .. r.range_end.to_u32)
+        case [existing_range.include?(range_begin_u32), existing_range.include?(range_end_u32)]
+        when [true, true]
+          # exsiting range includes given range.
+          # => this case can be ignored.
+        when [true, false]
+          # given begin address is in existing range but
+          # given end address is higher than existing range.
+          # existing: 192.168.0.10-20
+          # given: 192.168.0.19-21
+          # result: 192.168.0.10-21
+          # => expand r.range_end or merge
+          #    perform it after all iteration.
+          to_be_merged[0] = r
+        when [false, true]
+          # given begin address is lower than existing range but
+          # given end address is in existing range.
+          # => expand r.range_begin or merge
+          #    perform it after all iteration.
+          to_be_merged[1] = r
+        when [false, false]
+          # given range includes existing range.
+          # => this case sholud be handled at above ``included_ranges().destroy``
+          raise "BUG: Dead Code Path"
+        end
+      }
 
-      self
+      case to_be_merged.map(&:class)
+      when [NilClass, NilClass]
+        # no marge targets.
+        # Here comes when given range includes one or more exsisting
+        # ranges so that they are destroyed and add the given range as new.
+        # existing: 192.168.0.10-20, 30-40
+        # given: 192.168.0.9-41
+        # result: destroy => 10-20 & 30-40, add 9-41.
+        self.add_dhcp_range(range_begin: range_begin_u32,
+                            range_end: range_end_u32)
+      when [DhcpRange, NilClass]
+        # expand its end address
+        to_be_merged[0].tap { |r|
+          r.range_end = range_end_u32
+          r.save_changes
+        }
+      when [NilClass, DhcpRange]
+        # expand its end address
+        to_be_merged[1].tap { |r|
+          r.range_begin = range_begin_u32
+          r.save_changes
+        }
+      when [DhcpRange, DhcpRange]
+        # merge two ranges.
+        to_be_merged[0].tap { |r|
+          r.range_end = to_be_merged[1].range_end
+          r.save_changes
+        }
+        to_be_merged[1].destroy
+      else
+        raise "BUG: Dead Code Path"
```

Comment:
Same as the above. Making this code thread safe would make this impossible to occur and then it can be deleted.